### PR TITLE
Add full 'request' api to 'request-promise'

### PR DESCRIPTION
**Pull request by joeskeen**
According to the documentation for request-promise (https://github.com/request/request-promise):

> Since version 0.3.0 Request-Promise is not a wrapper around Request anymore. It now adds a .then(...) method to the Request prototype and exports the original Request object. This means you can now use all features of Request.

This PR brings the definitions for the full `request` API  to `request-promise`.

In the process of doing this, I added more tests to `request` from their project page (https://github.com/request/request) to bring it up to date, too.  I then used the same tests on `request-promise`, getting the idea from what I read at https://github.com/request/request-promise#can-i-trust-this-module:

> Additionally, the original tests of Request were executed on Request-Promise to ensure that we can call it "a drop-in replacement for Request".

As the changes are extensive, I would like some feedback from the original .d.ts authors:
- request.d.ts: @soywiz, @bonnici, or @Bartvds
- request-promise.d.ts: @cglantschnig or @AyaMorisawa

### CODE REVIEW

**Comment by joeskeen** [comment]
I added another commit because I read this at https://github.com/request/request-promise#promise---for-advanced-use-cases:
> In order to not pollute the Request call objects with the methods of the underlying Bluebird promise, only `.then(...)`, `.catch(...)`, and `.finally(...)` were exposed to cover most use cases. The effect is that any methods of a Bluebird promise other than `.then(...)`, `.catch(...)`, or `.finally(...)` cannot be used as the FIRST method in the promise chain

**Comment by vvakame** [comment]
*request-promise/request-promise.d.ts*

to author(@cglantschnig). could you review this PR?
:+1: or :-1:?

check list

* [ ] pass the Travic-CI test?

---

*request/request.d.ts*

to authors(@soywiz @bonnici @Bartvds). could you review this PR?
:+1: or :-1:?

check list

* [ ] pass the Travic-CI test?

**Comment by cglantschnig** [comment]
:+1: 
thanks for updating this library

**Comment by joeskeen** [comment]
@soywiz @bonnici or @Bartvds could you please review this?

**Comment by joeskeen** [comment]
@soywiz @bonnici or @Bartvds could you please review this?
@vvakame how long do we wait for a response?

**Comment by enriched** [comment]
Is the scope a bit too big on this pull request?  Would it be possible to split it into a pull request for the reqest library and one for the request-promise?

File: request/request.d.ts [hunk 1]
```diff
@@ -13,57 +13,54 @@ declare module 'request' {
 	import http = require('http');
 	import FormData = require('form-data');
 	import url = require('url');
+	import fs = require('fs');
 
-	export = RequestAPI;
+	namespace request {
+		export interface RequestAPI<TRequest extends Request, TOptions extends OptionalOptions> {
```
Comment (enriched) [inline request/request.d.ts]: Would it be better to leave the RequestAPI untemplated and then override the functions definitions in RequestPromiseAPI?  I definitely agree with moving this to a interface though, the module with function exports isn't very flexible.

**Comment by vvakame** [comment]
okay, i decided to merge this PR. thanks mate.

> scope a bit too big on this pull request

:+1:

**Pull request merged**